astropy/time/tests/test_basic.py

```diff
@@ -1168,3 +1167,138 @@ def test_bytes_input():
     assert tarray.dtype.kind == 'S'
     t2 = Time(tarray)
     assert t2 == t0
+
+
+def test_writeable_flag():
+    t = Time([1, 2, 3], format='cxcsec')
+    t[1] = 5.0
+    assert allclose_sec(t[1].value, 5.0)
+
+    t.writeable = False
+    with pytest.raises(ValueError) as err:
+        t[1] = 5.0
+    assert 'Time object is read-only. Make a copy()' in str(err)
+
+    with pytest.raises(ValueError) as err:
+        t[:] = 5.0
+    assert 'Time object is read-only. Make a copy()' in str(err)
+
+    t.writeable = True
+    t[1] = 10.0
+    assert allclose_sec(t[1].value, 10.0)
+
+    # Scalar is not writeable
+    t = Time('2000:001', scale='utc')
+    with pytest.raises(ValueError) as err:
+        t[()] = '2000:002'
+    assert 'scalar Time object is read-only.' in str(err)
+
+    # Transformed attribute is not writeable
+    t = Time(['2000:001', '2000:002'], scale='utc')
+    t2 = t.tt  # t2 is read-only now because t.tt is cached
+    with pytest.raises(ValueError) as err:
+        t2[0] = '2005:001'
+    assert 'Time object is read-only. Make a copy()' in str(err)
+
+
+def test_setitem_location():
+    loc = EarthLocation(x=[1, 2] * u.m, y=[3, 4] * u.m, z=[5, 6] * u.m)
+    t = Time([[1, 2], [3, 4]], format='cxcsec', location=loc)
+
+    # Succeeds because the right hand side makes no implication about
+    # location and just inherits t.location
+    t[0, 0] = 0
+    assert allclose_sec(t.value, [[0, 2], [3, 4]])
+
+    # Fails because the right hand side has location=None
+    with pytest.raises(ValueError) as err:
+        t[0, 0] = Time(-1, format='cxcsec')
+    assert 'cannot set to Time with different location' in str(err)
```

Comment:
The error message doesn't seem quite right here - isn't the issue more that on the right the location isn't set so it's ambiguous?

Reply:
It's different in the sense that the left side has `location = EarthLocation(...)` while the right side has `location = None`.  I think the exception will let the user diagnose the problem by hinting to print the location of left and right sides.  I could make the exception message be `.. different location attribute` to be even more clear?  Or something else?

Reply:
Yes, saying different location attribute - maybe could you explicitly say what the location is in each case? (expected ..., got ...)